A43-sdk-authorization.md

```diff
@@ -0,0 +1,364 @@
+A43: SDK authorization
+
+Author(s): [Ashitha Santhosh](https://github.com/ashithasantosh)
+Approver: 
+Status: Draft {Draft, In Review, Ready for Implementation, Implemented}
+Implemented in: <language, ...>
+Last updated: 2021-06-17
+Discussion at: (filled after thread exists)
+
+##Abstract
+
+gRPC is currently implementing an open-source standard authorization solution
+based on Envoy RBAC. This proposal discusses how this framework would be made 
+available to OSS gRPC users to perform per-RPC authorization checks.
+
+##Background
+
+gRPC authorization support which is currently under implementation, internally 
+implements RBAC Engines based on Envoy RBAC policies. RBAC policy provides 
+service-level and method-level access control for a service. Engines process 
+incoming RPC request attributes against policy configs and make a decision on 
+whether to allow or deny the request. The decision depends on the type of policy 
+(if the policy is an allowlist or denylist) and whether a matching policy was found.
+
+[RBAC filter]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/rbac/v3/rbac.proto
+
+The RBAC engine implementation is used by xDS API (i.e. Proxyless Service Mesh, 
+when Istio or Traffic Director sends Envoy RBAC policy to xDS enabled gRPC servers 
+for authorization) and SDK API (which makes authorization available to OSS gRPC users). 
+This document shares details on the latter.
+
+##Related Proposals:
+
+* [A41: xds RBAC Support][A41]
+
+##Proposal
+
+### User facing Authorization Policy
+
+gRPC SDK authorization policy is the user facing policy language that allows
+service owners and/or security admins to enable per-RPC authorization checks.
+
+SDK authorization policy guarantees stability and backward compatibility. So if
+a user writes a config, it will always be supported by gRPC unless the policy
+undergoes major versioning change like adding required fields. The policy is 
+extendable, we may add new fields in the future as new requirements are introduced, 
+we won’t be removing fields or updating semantics of existing fields.
+
+Following is the JSON schema of SDK Authorization Policy Version 1.0
+
+```json
+{
+	"title": "AuthorizationPolicy",
+	"definitions": {
+		"rule": {
+			"description": "Specification of rules. An empty rule is always matched (i.e., both source and request are empty)",
+			"type" : "object",
+			"properties": {
+				"name": {
+					"description": "The name of an authorization rule. It is mainly for monitoring and error message generation.",
+					"type": "string"
+				},
+				"source": {
+					"description": "Specifies attributes of a peer. Fields in the source are ANDed together, once we support" 
+					  "multiple fields in the future. If not set, no checks will be performed against the source.",
+					"type": "object",
+					"properties": {
```

Comment:
Only `principals` supported? What about source IP address? Why can't we support that? If mTLS is not set then principals cannot be used. Should we mention that?

Reply:
I included details on how we would match the principal. This also covers your suggestion, that if TLS is not used, then principals should fail to match.

In the SDK design document, we only have principals in source. I am guessing the fields were decided based on the current requirements (from Brixia), and source IP wasn't required. So to keep it simple, it wasn't included in MVP. 
Wdyt about adding source ip later, once we have MVP implemented?

Reply:
> Wdyt about adding source ip later, once we have MVP implemented?

Sounds good to me if that was the plan. But if the engine is going to be shared with the xDS RBAC implementation and that one does support source IP address, I wonder what's the additional effort to add here.

Reply:
I am not aware of any plan on adding source_ip to SDK policy. Maybe @markdroth @ejona86 @dfawley would know. 

About the shared engine implementation. There are several fields in RBAC that aren't supported by SDK authorization policy. It does not affect the SDK path, as we would be translating the SDK policy to Envoy RBAC (which would only contain the SDK fields), and the engine evaluation just works with that policy. So there is nothing extra that we have to do in the engine, to support SDK authorization.

Translator code
https://github.com/grpc/grpc/blob/master/src/core/lib/security/authorization/rbac_translator.h#L34

Reply:
> ... So there is nothing extra that we have to do in the engine, to support SDK authorization.

Exactly. So my point is if the shared engine supports (source) IP address in the Envoy RBAC policy, the SDK authz feature can take advantage of that with minimal efforts for adding source IP support unless my assumptions are wrong...